bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -176,5 +207,163 @@ def to_parquet(dataframe, bq_schema, filepath):
     if pyarrow is None:
         raise ValueError("pyarrow is required for BigQuery schema conversion.")
 
-    arrow_table = to_arrow(dataframe, bq_schema)
+    arrow_table = dataframe_to_arrow(dataframe, bq_schema)
     pyarrow.parquet.write_table(arrow_table, filepath)
+
+
+def _tabledata_list_page_to_dataframe(page, column_names, dtypes):
+    columns = collections.defaultdict(list)
+    for row in page:
+        for column in column_names:
+            columns[column].append(row[column])
+    for column in dtypes:
+        columns[column] = pandas.Series(columns[column], dtype=dtypes[column])
+    return pandas.DataFrame(columns, columns=column_names)
+
+
+def download_dataframe_tabledata_list(pages, schema, dtypes):
+    """Use (slower, but free) tabledata.list to construct a DataFrame."""
+    column_names = [field.name for field in schema]
+    for page in pages:
+        yield _tabledata_list_page_to_dataframe(page, column_names, dtypes)
+
+
+def _download_dataframe_bqstorage_stream(
+    download_state,
+    bqstorage_client,
+    column_names,
+    dtypes,
+    session,
+    stream,
+    worker_queue,
+):
+    position = bigquery_storage_v1beta1.types.StreamPosition(stream=stream)
+    rowstream = bqstorage_client.read_rows(position).rows(session)
+
+    for page in rowstream.pages:
+        if download_state.done:
+            return
```

Comment:
There are two ways that `download_state.done` could be set:

1. All streams have completed and been processed. (This branch won't be hit in that case, as the thread is already exited.)
2. One of the worker threads has thrown an exception (or KeyboardInterrupt or something was thrown in the main thread). In this case, we don't care at all about in-flight frames because everything is getting thrown out anyway.